contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,654 @@
+for using raw block devices, the legacy method of populating the devices and appending it to the device array is sufficient.
+
+# Implementation Plan, Features & Milestones
+
+Phase 1: v1.8
+Feature: Pre-provisioned PVs to precreated devices 
+
+               Milestone 1: API changes
+
+               Milestone 2: Restricted Access 
+              
+               Milestone 3: Changes to the mounter interface as today it is assumed 'file' as the default.
+               
+               Milestone 4: Expose volumeMode to users via kubectl
+               
+               Milestone 5: Adds enable/disable configuration to securityContext in PSP (Pod Security Policy) similar to hostPath
+               
+               Milestone 6: Validate container runtime options with user specifcations as indicated in UC3
+	       
+	       Milestone 7: Container Runtime changes
+
+Phase 2:  v1.9
+Feature: Discovery of block devices 
+
+                Milestone 1: Dynamically provisioned PVs to dynamically allocated devices
+
+                Milestone 2: Privileged container concerns    
+		
+		Milestone 3: Flex volume updates
+
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+## Mounter interface proposed design
+# Plugin changes
+## New BlockVolumeMapper interface proposed design
+
+```
+ type BlockVolumeMapper interface {
+       Volume
+       CanBlockMap() error
+       SetUpDevice() error
+       SetUpDeviceAt(dir string) error
+       GetVolumePath() string
+ }
+ type BlockVolumeUnmapper interface {
+       Volume
+       TearDownDevice() error
+       TearDownDeviceAt(dir string) error
+ }
+```
+## Changes for volume mount points
+
+Currently, a volume which has filesystem is mounted to the following two paths on a kubelet node when the volumes is in-use.
+The purpose of those mount points are that Kubernetes manages volume attach/detach status using these mount points and number
+of references to these mount points.
+
+```
+- Global mount path
+`/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/{volumePluginDependentPath}/`
+
+- Volume mount path
+`/var/lib/kubelet/pods/{podUID}/volumes/{escapeQualifiedPluginName}/{volumeName}/`
+```
+
+Even if the volumeMode is "Block", similar scheme is needed. However, the volume which 
+doesn't have filesystem can't be mounted.
+Therefore, instead of volume mount, we use symbolic link.
+- Kubelet creates a new symbolic link under the new global mount path when volume is attached to a Pod. Number of symbolic links
+is equal to the number of Pods which uses the same volume. Kubelet needs to manage both creation and deletion of symbolic links
+under the global mount path.
+ 
+```
+Global mount path for "Block" mode volume
+/var/lib/kubelet/plugins/kubernetes.io/{pluginName}/volumeDevices/{volumePluginDependentPath}/
```

Comment:
Could the name `volumeDevices` conflict with the `volumePluginDependentPath`?

Reply:
@msau42 you mean, could they be the same and that be a problem? I don't understand the conflict

Reply:
Nm, I was thinking the name could conflict with the filesystem volumes path, but I think it's ok.

The filesystem volumes plugin directory is:
/var/lib/kubelet/plugins/<plugin name>/mounts/<volume specific name>

For block volumes, `mounts` is replaced with `volumeDevices`

Reply:
@msau42 correct